### PR TITLE
Feature/mlt 29 as a user i want to be able to register as a new user

### PR DESCRIPTION
**Pull request by imrankhan-stixor**
# Description

Added the signup UI, api calls for sign up, storing of data in redux, UI for verification, functionality and api calls for the verification

## Type of change

Please select type of issue addressed,

- [x] New feature ( MLT-29 ) 

# Dependences / Libraries Installed

Please list the packages and libraries installed (if any)

- [x] ( @radix-ui/react-icons )  ( added by shadcn ) 

# How Has This Been Tested By The Developer?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tested the api calls to work correctly
- [x] Tested the handling of the api responses.
- [x] Tested the Ui for responsiveness.
- [x] Tested the redirections to work properly
- [x] Tested the results to match the expected outcome

# Checklist:

- [x] I have performed a self review of my code.

- [x] I have removed consoles, debuggers and unnecessary code blocks.

- [x] My code follows the style guidelines of this project

- [x] My changes generate no new warnings or errors

- [ ] I have commented my code, particularly in hard-to-understand areas

- [ ] New and existing unit tests pass locally with my changes

- [x] Any dependent changes have been merged and published in downstream modules

- [x] I understand that I am accountable for any code added in this PR, and any negative impact will effect my performance evaluation

### CODE REVIEW

File: components/ui/button.tsx [hunk 1]
```diff
@@ -33,17 +37,29 @@ export interface ButtonProps
...
       <Comp
-        className={cn(buttonVariants({ variant, size, className }))}
+        className={cn(
+          buttonVariants({ variant, size, className }),
+          loading && "relative",
+        )}
         ref={ref}
         {...props}
-      />
+      >
```
Comment (fawadmehmood1) [inline components/ui/button.tsx:57]: ` {loading ? <ReloadIcon className="mr-2 h-4 w-4 animate-spin" /> :  <span>{props.children}</span>}`

File: components/ui/input.tsx [hunk 2]
```diff
@@ -1,25 +1,48 @@
...
-      <input
-        type={type}
-        className={cn(
-          "flex h-10 w-full rounded-md border border-inputBorder bg-background px-3 py-2 text-sm text-placeholder outline-none file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-sm placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50 lg:w-[40rem]",
-          className,
+      <div className="relative">
+        <input
+          type={(isPassword && !showPassword) || !isPassword ? type : "text"}
+          className={cn(
+            "!mt-[0.375rem] flex h-10 w-full rounded-md border border-inputBorder bg-background px-3 py-2 text-sm text-placeholder outline-none file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-sm placeholder:text-muted-foreground focus-visible:outline-none disabled:cursor-not-allowed disabled:opacity-50 sm:w-[70vw] lg:w-[39.5rem]",
```
Comment (fawadmehmood1) [inline components/ui/input.tsx]: never do `focus-visible:outline-none`, it removes focus indication for keyboard users. You can use you `primary` color here.

File: components/ui/button.tsx [hunk 3]
```diff
@@ -3,27 +3,31 @@ import { Slot } from "@radix-ui/react-slot";
 import { cva, type VariantProps } from "class-variance-authority";
 
 import { cn } from "@/lib/utils";
+import { ReloadIcon } from "@radix-ui/react-icons";
```
Comment (fawadmehmood1) [inline components/ui/button.tsx]: we are already using `lucide-react`, use this instead `import { Loader2 } from "lucide-react"`
Remove this library

File: package.json [hunk 4]
```diff
@@ -20,6 +20,7 @@
   "dependencies": {
     "@hookform/resolvers": "^3.3.4",
     "@radix-ui/react-dialog": "^1.0.5",
+    "@radix-ui/react-icons": "^1.3.0",
```
Comment (fawadmehmood1) [inline package.json]: Remove this library

File: redux/apis/api-types.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+export interface ResendEmailVerificationLinkResponse {
+  data: {
+    message: string;
+    data: string;
+  };
+}
+
+export interface VerifyEmailRequest {
```
Comment (fawadmehmood1) [inline redux/apis/api-types.ts]: You can directly use a string where you are currently using this type. There isn't much available here

File: redux/apis/auth-api.ts [hunk 6]
```diff
@@ -19,7 +27,42 @@ export const authApi = createApi({
...
+      ResendEmailVerificationLinkRequest
+    >({
+      query: (body) => ({
+        method: "POST",
+        url: "/user/send-verification-link",
+        body: body,
+      }),
+    }),
+
+    verifyLink: builder.query<any | ErrorResponse, VerifyEmailRequest>({
```
Comment (fawadmehmood1) [inline redux/apis/auth-api.ts]: you can directly use `string` here

File: redux/apis/auth-api.ts [hunk 7]
```diff
@@ -19,7 +27,42 @@ export const authApi = createApi({
   tagTypes: [],
 
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (fawadmehmood1) [inline redux/apis/auth-api.ts]: remove this

File: types/index.ts [hunk 8]
```diff
@@ -1,6 +1,30 @@
+import { LucideProps } from "lucide-react";
+import { ForwardRefExoticComponent, RefAttributes } from "react";
+
+export interface RootState {
+  user: UserState;
+}
+
 export interface UserState {
   user?: {
```
Comment (fawadmehmood1) [inline types/index.ts]: redundant interface, this is already defined in `VerifyEmailResponse`

**Review by fawadmehmood1 — CHANGES_REQUESTED** [review]

File: app/(auth)/verification/page.tsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  }, [data]);
+
+  const handleClick = async () => {
+    if (!token) {
+      try {
+        await resendLink({
+          email: storedEmail as string,
+          local: LOCAL,
+        }).unwrap();
+        alert("Link sent successfully");
```
Comment (hassantahir767) [inline app/(auth)/verification/page.tsx]: The message should be more meaningful, should convey message properly.

File: components/social-sign-up.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+const SocialSignUp = ({ title = "signup" }: { title?: string }) => {
+  const socialLink = (icon: ReactNode, name: string) => {
+    return (
+      <Button
+        className="flex grow rounded-md border px-4 py-2"
+        variant={"transparent"}
+      >
+        <span className="flex items-center justify-center gap-[0.5rem]">
+          {" "}
```
Comment (hassantahir767) [inline components/social-sign-up.tsx]: remove {" "} if not required

File: components/social-sign-up.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            width={18}
+            height={18}
+          />,
+          "Google",
+        )}
+
+        {socialLink(
+          <Image
+            src={"/assets/images/social-icons/slack-icon.svg"}
+            alt="google icon"
```
Comment (hassantahir767) [inline components/social-sign-up.tsx]: `alt = "slack"`

File: components/social-sign-up.tsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            width={18}
+            height={18}
+          />,
+          "Slack",
+        )}
+
+        {socialLink(
+          <Image
+            src={"/assets/images/social-icons/jira-icon.svg"}
+            alt="google icon"
```
Comment (hassantahir767) [inline components/social-sign-up.tsx]: `alt = "jira"`

File: components/ui/input.tsx [hunk 13]
```diff
@@ -1,25 +1,48 @@
...
 export interface InputProps
-  extends React.InputHTMLAttributes<HTMLInputElement> {}
+  extends React.InputHTMLAttributes<HTMLInputElement> {
+  isPassword?: boolean;
+}
 
 const Input = React.forwardRef<HTMLInputElement, InputProps>(
-  ({ className, type, ...props }, ref) => {
+  ({ className, type, isPassword = false, ...props }, ref) => {
+    const [showPassword, setShowPassword] = React.useState(false);
```
Comment (hassantahir767) [inline components/ui/input.tsx]: import useState from react above. That way you won't need to write React.useState each time. 

Also, use useState<boolean>(false);

**Review by hassantahir767 — CHANGES_REQUESTED** [review]

**Review by fawadmehmood1 — APPROVED** [review]

**Review by hassantahir767 — APPROVED** [review]

**Pull request merged**